// serialist/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.exception = exports.SerializedTypes = void 0;
const deserialize_1 = require("./deserialize");
const serialize_1 = require("./serialize");
var SerializedTypes;
(function (SerializedTypes) {
    SerializedTypes[SerializedTypes["Object"] = 0] = "Object";
    SerializedTypes[SerializedTypes["Special"] = 1] = "Special";
    SerializedTypes[SerializedTypes["Constructed"] = 2] = "Constructed";
    SerializedTypes[SerializedTypes["Custom"] = 3] = "Custom";
    SerializedTypes[SerializedTypes["Array"] = 4] = "Array";
    SerializedTypes[SerializedTypes["Reference"] = 5] = "Reference";
})(SerializedTypes = exports.SerializedTypes || (exports.SerializedTypes = {}));
function exception(message) {
    throw new Error(message);
}
exports.exception = exception;
function default_1(constructors, options) {
    const fnSerialize = serialize_1.default(constructors || [], options || {});
    const fnDeserialize = deserialize_1.default(constructors || [], options || {});
    return {
        deserialize: fnDeserialize,
        parse: (stringified) => fnDeserialize(JSON.parse(stringified)),
        serialize: fnSerialize,
        stringify: (source) => JSON.stringify(fnSerialize(source))
    };
}
exports.default = default_1;
